library/core/src/iter/adapters/zip.rs

```diff
@@ -171,26 +246,109 @@ macro_rules! zip_impl_general_defaults {
             // and doesn’t call `next_back` too often, so this implementation is safe in
             // the `TrustedRandomAccessNoCoerce` specialization
 
-            let a_sz = self.a.len();
-            let b_sz = self.b.len();
-            if a_sz != b_sz {
-                // Adjust a, b to equal length
-                if a_sz > b_sz {
-                    for _ in 0..a_sz - b_sz {
-                        self.a.next_back();
-                    }
-                } else {
-                    for _ in 0..b_sz - a_sz {
-                        self.b.next_back();
-                    }
-                }
-            }
+            self.adjust_back();
             match (self.a.next_back(), self.b.next_back()) {
                 (Some(x), Some(y)) => Some((x, y)),
                 (None, None) => None,
                 _ => unreachable!(),
             }
         }
+
+        #[inline]
+        default fn fold<T, F>(self, init: T, mut f: F) -> T
+        where
+            F: FnMut(T, Self::Item) -> T,
+        {
+            let mut a = self.a;
+            let mut b = self.b;
+
+            let acc = a.try_fold(init, move |acc, x| match b.next() {
+                Some(y) => Ok(f(acc, (x, y))),
+                None => Err(acc),
+            });
+
+            match acc {
+                Ok(exhausted_a) => exhausted_a,
+                Err(exhausted_b) => exhausted_b,
+            }
+        }
+
+        #[inline]
+        default fn try_fold<T, F, R>(&mut self, init: T, mut f: F) -> R
+        where
+            F: FnMut(T, Self::Item) -> R,
+            R: Try<Output = T>,
+        {
+            let a = &mut self.a;
+            let b = &mut self.b;
+
+            let acc = a.try_fold(init, move |acc, x| match b.next() {
+                Some(y) => {
+                    let result = f(acc, (x, y));
+                    match result.branch() {
+                        ControlFlow::Continue(continue_) => Ok(continue_),
+                        ControlFlow::Break(break_) => Err(R::from_residual(break_)),
+                    }
+                }
+                None => Err(R::from_output(acc)),
+            });
+
+            match acc {
+                Ok(ok) => R::from_output(ok),
+                Err(err) => err,
+            }
+        }
+
+        #[inline]
+        default fn rfold<T, F>(mut self, init: T, mut f: F) -> T
+        where
+            A: DoubleEndedIterator + ExactSizeIterator,
+            B: DoubleEndedIterator + ExactSizeIterator,
+            F: FnMut(T, Self::Item) -> T,
+        {
+            self.adjust_back();
+            let mut a = self.a;
+            let mut b = self.b;
+
+            let acc = a.try_rfold(init, move |acc, x| match b.next_back() {
+                Some(y) => Ok(f(acc, (x, y))),
+                None => Err(acc),
+            });
+
+            match acc {
+                Ok(exhausted_a) => exhausted_a,
+                Err(exhausted_b) => exhausted_b,
+            }
+        }
```

Comment:
Here we know that the inner iterators have the same length (after adjusting), so it might be worth writing this in terms of `self.a.rfold` instead. We'd have to panic in case `b.next_back` returns `None`. I have not benchmarked this, but in theory `rfold` might lead to more efficient machine code, depending on the underlying iterator.

Reply:
this gave me the idea of applying a similar optimization to `fold`. by checking if both iterators have the same length (which i assumed is a common case for `zip`), we can use `a.fold` instead of `a.try_fold`

Reply:
Very cool! I didn't think about doing that with `size_hint`.